### DISCUSSION TITLE
Configuration problem

### DISCUSSION THREAD

**Discussion by timjs** [gleam-lang/gleam#1284] [Ideas & suggestions]
(As a follow up on #1244 and #1272.)

The configuration problem was once stated by [Oleg Kiselyov in _Functional Pearl: Implicit Configurations_](http://homes.sice.indiana.edu/ccshan/prepose/prepose.pdf). It's best described by an example:

```rust
type Modulo { Modulo(k: Int) };

fn add(a: Int, b: Int, mod: Modulo) {
  a + b % mod.k
}

fn mul(a, b, mod) {
  a * b % mod.k
}

fn test(a, b, mod) {
  add(mul(a, a, mod), mul(b, b, mod), mod)
}

fn main() {
  let mod = Modulo(4)
  print(test(2, 3, mod))  // Explicitly pass `Modulo(4)` as argument

  let mod = Modulo(5)
  print(test(2, 3, mod))  // Explicitly pass `Modulo(5)` as argument
}
```

Problem here is that the `mod` argument has to be passed explicitly to every function call. It is similar, but not equivalent, to pass around configuration data using a Reader monad in Haskell/PureScript/Idris, as the modulo problem show above _abstracts over the current context_.

Implicts as used by Scala are a possible solution to this problem. Question is: do we like to address this problem in Gleam, i.e. is it something that happens a lot, or is it a niche and is explicit passing not a big problem?

**Comment by lpil** [comment]
For me I think this is a more appealling argument in favour of implicits than polymorphism.

I do not mind the slightly more verbose APIs of monomorphic functions, and I think there is also value in being able to quickly identify where a function is implement or what data structures are being used purely by looking at function names.

I cannot on the other hand see much value in having to pass an instance of `DatabaseConnection` to every query and `LoggerMetadata` to every log call. I think this would be mildly irritating and at times make the code noisier to read.

**Comment by lpil** [comment]
Hi again! I was wondering @timjs if you knew of other solutions to this problem? Ignoring ad-hoc polymorphism for the time being. I'm aware of other approaches to polymorphism but not this.

**Reply by timjs** [reply]
I'm aware of three:

1. Implicit arguments
2. The reader monad
3. (Algebraic) effects

Can't do a longer write up atm, maybe at the end of the weekend.

**Reply by lpil** [reply]
Ah yes, I didn't think of those. Thanks!

**Reply by timjs** [reply]
Is there a medium sized real world example I could use to experiment with and show similarities and differences of each approach in a write up?

**Reply by lpil** [reply]
I don't think there is a public one I'm afraid
